components/tls/tls-client/src/kx.rs

```diff
@@ -50,8 +50,18 @@ impl KeyExchange {
         f: impl FnOnce(&[u8]) -> Result<T, ()>,
     ) -> Result<T, Error> {
         let peer_key = ring::agreement::UnparsedPublicKey::new(self.skxg.agreement_algorithm, peer);
-        ring::agreement::agree_ephemeral(self.privkey, &peer_key, (), f)
-            .map_err(|()| Error::PeerMisbehavedError("key agreement failed".to_string()))
+        let ret = ring::agreement::agree_ephemeral(self.privkey, &peer_key, f);
+        match ret {
+            Ok(ok_value) => match ok_value {
+                Ok(t) => Ok(t),
+                Err(()) => Err(Error::PeerMisbehavedError(
+                    "key agreement failed".to_string(),
+                )),
+            },
+            Err(_) => Err(Error::PeerMisbehavedError(
```

Comment:
Thanks for the review! The original
```rust
        ring::agreement::agree_ephemeral(self.privkey, &peer_key, (), f)
            .map_err(|()| Error::PeerMisbehavedError("key agreement failed".to_string()))
```
fails because `f` returns `Result<T, ()>` and thus [`ring::agreement::agree_ephemeral`](https://github.com/briansmith/ring/blob/6347be9634676761dcedf63f2ea48f866f798a16/src/agreement.rs#L245-L249) would return `Result<Result<T, ()>, Unspecified>`. To make it work, it seems we would need an extra `unwrap` to make the return value become the type `Result<T, Error>`? Like the following code snippet

```rust
        ring::agreement::agree_ephemeral(self.privkey, &peer_key, f)
            .unwrap().map_err(|_| Error::PeerMisbehavedError("key agreement failed".to_string()))
```
or should we change the type of `f` from `impl FnOnce(&[u8]) -> Result<T, ()>` to `impl FnOnce(&[u8]) -> T`? or am I understanding it wrong?

Reply:
I think it is the last thing you mentioned:
```rust
    pub(crate) fn complete<T>(self, peer: &[u8], f: impl FnOnce(&[u8]) -> T) -> Result<T, Error> {
        let peer_key = ring::agreement::UnparsedPublicKey::new(self.skxg.agreement_algorithm, peer);
        ring::agreement::agree_ephemeral(self.privkey, &peer_key, f)
            .map_err(|_| Error::PeerMisbehavedError("key agreement failed".to_string()))
    }
```

Reply:
Thanks for the direction! Changed it in https://github.com/tlsnotary/tlsn/pull/391/commits/6c597d349e2b9f8a9faf1d7632594a878b4cc5e4